tutorials/crunch-aggregate-sessions.md

```diff
@@ -0,0 +1,296 @@
+
+## Understanding the AggregateEvents Class
+
+AggregateEvents uses a Crunch `Pipeline` object to apply functions to input data. The general workflow in Crunch is:
+
+1. Read input (for example, a dataset) into a Pipeline object.
+1. Perform functions on the input data.
+1. Write the result to a destination (for example, another dataset).
+
+The following is the complete code listing for AggregateEvents, with comments inline.
+
+### AggregateEvents.java
+
+```Java
+package org.kitesdk.examples.tutorials.crunch;
+
+import java.io.Serializable;
+import java.net.URI;
+import java.util.Calendar;
+import java.util.Iterator;
+import java.util.TimeZone;
+import org.apache.crunch.DoFn;
+import org.apache.crunch.Emitter;
+import org.apache.crunch.MapFn;
+import org.apache.crunch.PCollection;
+import org.apache.crunch.Pair;
+import org.apache.crunch.Target;
+import org.apache.crunch.types.avro.Avros;
+import org.apache.crunch.util.CrunchTool;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.util.ToolRunner;
+import org.kitesdk.data.Dataset;
+import org.kitesdk.data.Datasets;
+```
+Kite's CrunchDatasets class methods handle the underlying code required to create source and target datasets for use with Crunch.
+
+```Java
+import org.kitesdk.data.crunch.CrunchDatasets;
+import org.kitesdk.data.event.StandardEvent;
+import org.kitesdk.data.spi.filesystem.FileSystemDatasets;
+import org.kitesdk.examples.demo.event.Session;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AggregateEvents extends CrunchTool implements Serializable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AggregateEvents.class);
+
+  @Override
+  public int run(String[] args) throws Exception {
+
+    // Turn debug on while in development.
+    getPipeline().enableDebug();
+    getPipeline().getConfiguration().set("crunch.log.job.progress", "true");
+
+```
+
+Step 1. Load the source dataset. In this example, use the `events` dataset from Hive.
+
+```Java
+    Dataset<StandardEvent> eventsDataset = Datasets.load(
+        "dataset:hive:events", StandardEvent.class);
+```
+
+If the dataset is empty, stop.
+
+```Java
+    if (eventsDataset.isEmpty()) {
+      LOG.info("No records to process.");
+      return 0;
+    }
+```
+
+Still going? Good. Create a _parallel collection_ using the `events` dataset as the source. A parallel collection (`PCollection`) is a set of distributed objects that can be processed in parallel.
+
+```Java
+    PCollection<StandardEvent> events = read(
+        CrunchDatasets.asSource(eventsDataset));
+```
+
+Step 2. Process the data.
+a. Create a session key by combining the user ID and session ID.
+b. Group together all events with the same session key.
+c. For each group, create a Session record as an Avro Specific object. 
```

Comment:
I think the bit about the Avro Specific object is distracting. The main idea is to create a session record. If you think it is valuable to point out that it is a specific object (which I would agree with) then it can be done later in discussion.